File: backend/auth/middlewares/logging.go
package middlewares

import (
	"net"
	"net/http"
	"sen1or/letslive/auth/pkg/logger"
	"time"
)

type loggingResponseWriter struct {
	w          http.ResponseWriter
	statusCode int
	bytes      int
}

func (lrw *loggingResponseWriter) Header() http.Header {
	return lrw.w.Header()
}

func (lrw *loggingResponseWriter) Write(data []byte) (int, error) {
	wb, err := lrw.w.Write(data)
	lrw.bytes += wb
	return wb, err
}

func (lrw *loggingResponseWriter) WriteHeader(statusCode int) {
	lrw.w.WriteHeader(statusCode)
	lrw.statusCode = statusCode
}

func LoggingMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		timeStart := time.Now()
		lrw := &loggingResponseWriter{w: w}
		next.ServeHTTP(lrw, r)

		duration := time.Since(timeStart).Milliseconds()
		remoteAddr := r.Header.Get("X-Forwarded-For")
		if remoteAddr == "" {
			if ip, _, err := net.SplitHostPort(r.RemoteAddr); err != nil {
				remoteAddr = "unknown address"
			} else {
				remoteAddr = ip
			}
		}

		requestId := r.Context().Value("requestId")

		fields := []any{
			"requestId", requestId,
			"duration", duration,
			"method", r.Method,
			"remote#addr", remoteAddr,
			"response#bytes", lrw.bytes,
			"response#status", lrw.statusCode,
			"uri", r.RequestURI,
		}

		if lrw.statusCode/100 == 2 {
			if r.RequestURI != "/v1/health" {
				logger.Infow("success api call", fields...)
			}
		} else {
			err := lrw.w.Header().Get("X-LetsLive-Error")
			if len(err) == 0 {
				logger.Infow("failed api call", fields...)
			} else {
				logger.Errorw("failed api call: "+err, fields...)
			}
		}
	})
}

File: backend/chat/src/middlewares/logging.ts
import { Request, Response, NextFunction } from 'express'

function loggingMiddleware(req: Request, res: Response, next: NextFunction): void {
  const startTime = Date.now()
  const originalWrite = res.write.bind(res)
  const originalEnd = res.end.bind(res)

  let responseBytes = 0

  res.write = ((chunk: any, ...args: any[]) => {
    if (chunk) responseBytes += Buffer.byteLength(chunk)
    return originalWrite(chunk, ...args)
  }) as typeof res.write

  res.end = ((chunk: any, ...args: any[]) => {
    if (chunk) responseBytes += Buffer.byteLength(chunk)
    return originalEnd(chunk, ...args)
  }) as typeof res.end

  res.on('finish', () => {
    const duration = Date.now() - startTime
    const ip =
      req.headers['x-forwarded-for']?.toString().split(',')[0].trim() ||
      req.socket?.remoteAddress ||
      'unknown address'

    const requestId = (req as any).requestId || req.headers['x-request-id'] || 'unknown'

    const fields = {
      requestId,
      duration,
      method: req.method,
      'remote#addr': ip,
      'response#bytes': responseBytes,
      'response#status': res.statusCode,
      uri: req.originalUrl,
    }

    const isHealthCheck = req.originalUrl === '/v1/health'
    const letsLiveError = res.getHeader('X-LetsLive-Error') as string | undefined

    if (res.statusCode >= 200 && res.statusCode < 300) {
      if (!isHealthCheck) {
        console.info('success api call', fields)
      }
    } else {
      if (letsLiveError) {
        console.error(`failed api call: ${letsLiveError}`, fields)
      } else {
        console.info('failed api call', fields)
      }
    }
  })

  next()
}

export default loggingMiddleware;
